File: src/bmsx/machine/devices/audio/queue_status_registers.ts
import {
	IO_APU_CMD_CAPACITY,
	IO_APU_CMD_FREE,
	IO_APU_CMD_QUEUED,
	IO_APU_OUTPUT_CAPACITY_FRAMES,
	IO_APU_OUTPUT_FREE_FRAMES,
	IO_APU_OUTPUT_QUEUED_FRAMES,
} from '../../bus/io';
import type { ApuCommandFifo } from './command_fifo';
import type { ApuOutputRing } from './output_ring';

export class ApuQueueStatusRegisters {
	public constructor(
		private readonly commandFifo: ApuCommandFifo,
		private readonly outputRing: ApuOutputRing,
	) {}

	public read(addr: number): number {
		switch (addr) {
			case IO_APU_OUTPUT_QUEUED_FRAMES:
				return this.outputRing.queuedFrames();
			case IO_APU_OUTPUT_FREE_FRAMES:
				return this.outputRing.freeFrames();
			case IO_APU_OUTPUT_CAPACITY_FRAMES:
				return this.outputRing.capacityFrames();
			case IO_APU_CMD_QUEUED:
				return this.commandFifo.count;
			case IO_APU_CMD_FREE:
				return this.commandFifo.free;
			case IO_APU_CMD_CAPACITY:
				return this.commandFifo.capacity;
		}
		throw new Error('[APU] Queue-status register read was mapped to an unknown address.');
	}
}

File: src/bmsx_cpp/machine/devices/audio/queue_status_registers.cpp
#include "machine/devices/audio/queue_status_registers.h"

#include "machine/bus/io.h"
#include "machine/devices/audio/command_fifo.h"
#include "machine/devices/audio/output_ring.h"

namespace bmsx {

ApuQueueStatusRegisters::ApuQueueStatusRegisters(const ApuCommandFifo& commandFifo, const ApuOutputRing& outputRing)
	: m_commandFifo(commandFifo)
	, m_outputRing(outputRing) {}

u32 ApuQueueStatusRegisters::read(u32 addr) const {
	switch (addr) {
		case IO_APU_OUTPUT_QUEUED_FRAMES:
			return static_cast<u32>(m_outputRing.queuedFrames());
		case IO_APU_OUTPUT_FREE_FRAMES:
			return static_cast<u32>(m_outputRing.freeFrames());
		case IO_APU_OUTPUT_CAPACITY_FRAMES:
			return static_cast<u32>(m_outputRing.capacityFrames());
		case IO_APU_CMD_QUEUED:
			return m_commandFifo.count();
		case IO_APU_CMD_FREE:
			return m_commandFifo.free();
		case IO_APU_CMD_CAPACITY:
			return m_commandFifo.capacity();
	}
	throw BMSX_RUNTIME_ERROR("[APU] Queue-status register read was mapped to an unknown address.");
}

Value ApuQueueStatusRegisters::readThunk(void* context, u32 addr) {
	return valueNumber(static_cast<double>(static_cast<ApuQueueStatusRegisters*>(context)->read(addr)));
}

} // namespace bmsx
